### ISSUE TITLE
API: selective peer replication

### ISSUE THREAD

**Issue by noffle** [noffle/multifeed#2]
I'd really like APIs that offer more control around feeds:

- Drop (delete) a remote or local feed
- Have a userland callback at replication time that lets the user decide if they want to replicate a certain feed that a remote peer is offering

cc @0fork, who I was talking to about APIs to better support userland moderation and authorization models

**Comment by 0fork** [comment]
What would this require from the implementation perspective?

I would guess you could do a simple block list to drop feeds so that those are skipped in replication and user gets to choose whether they are permanently deleted or simply hidden. I can imagine a scenario where you don't want to actively replicate a feed anymore but it still has contents you would maybe like to access later.

**Comment by noffle** [comment]
Dropping a feed: right now feed loading is done by looking for the `key` file in each feed's storage. If you deleted that entry and also removed the feed from `this._feeds` that should do the trick.

Controlling feeds to replicate: the first thing that happens at replication time is both sides exchange a list of all hypercore keys they have & want to share. Replication could pause here, and there could be an API that gets called out to with the list of feeds, and returns the keys it wants to receive. There could be a similar API for choosing what feeds you'd like to share.

Dropping messages within a feed: we get this for free! hypercore already has a [clear](https://github.com/mafintosh/hypercore#feedclearstart-end-callback) API. I'm not sure if it will just automatically re-download it though the next time it connects to a peer with it. (cc @mafintosh on this)

**Comment by 0fork** [comment]
So to summarize, great news is that there’s nothing here that isn’t in the realm of just sitting down and doing the logic code and convenience functions around the logic.

Regarding dropping messages, isn’t redownloading avoidable by using sparse replication?

**Comment by noffle** [comment]
That all sounds about right. API design might be the hardest part!

**Comment by cblgh** [comment]
@0fork also check out this list of actions that we'd want to be able to do in cabal (a chat client making use of multifeed)

https://github.com/cabal-club/cabal/issues/41

**Comment by saranrapjs** [comment]
I've done some experiments with this locally, and it seems like with `sparse: true`, replication won't automatically download anything; e.g. if you `clear` a range, but want to replicate everything outside of that range,  you'd be on the hook for manually downloading everything _but_ that range by some means other than `replicate`. I do like the simplicity of downstream multifeed consumers being able to simply `replicate({ live: true})`...

Part of me wishes there was actually a hook in hypercore's replication code that allowed for ignoring offered hypercore blocks, based on some user-land logic. I for one have been looking into a way of specifying something like a `maxFeedLength` for the underlying multifeed hypercores, which could leverage something like this if it existed! You could also ignore specific messages this way, though maybe it'd be cumbersome

**Comment by noffle** [comment]
> Part of me wishes there was actually a hook in hypercore's replication code that allowed for ignoring offered hypercore blocks, based on some user-land logic.

I bet @mafintosh would be open to that!

**Comment by noffle** [comment]
I think @substack is working on this to implement peer blocks in Cabal.

**Comment by Juul** [comment]
Yeah I'm having to use multiplex and have multiple instances of multifeed to deal with different levels of access for different peers so something like this would be really great.
